### ISSUE TITLE
RFC: Test with ProjectBuilder instead of Gradle plugin "dry run"

### ISSUE THREAD

**Issue by 3flex** [detekt/detekt#3778]
## Expected Behavior
Tests are as fast as they can be.

## Current Behavior
The "dry run" tests aren't integration tests but use Gradle's TestKit, which is slow. The `:detekt-gradle-plugin:test` task takes longer to execute than any other task in the build.

## Context
Gradle's [ProjectBuilder](https://docs.gradle.org/current/javadoc/org/gradle/testfixtures/ProjectBuilder.html) is a speedier alternative to TestKit and can be used for tests that don't execute Gradle tasks.

The "dry run" tests are just checking that when certain options are set on the plugin that the flags passed to the CLI are as expected. A slight refactoring in the plugin combined with switching to ProjectBuilder for the dry run tests will speed things up a lot - individual tests that take 25-30 seconds with TestKit take 15 with ProjectBuilder on my machine.

The test results should also be picked up by JaCoCo so we can measure code coverage in the plugin.

An example (kotlin-dsl was used as the DSL can be used in the `project.run {}` block):

```kotlin
        it("does the test") {
            val project = ProjectBuilder.builder().build()

            project.run {
                apply<KotlinPluginWrapper>()
                apply<DetektPlugin>()
                repositories {
                    mavenCentral()
                    mavenLocal()
                }
                configure<DetektExtension> {
                    allRules = true
                    reports.txt.enabled = false
                }
            }

            project.getTasksByName("detektMain", true) // trigger project evaluation

            val detektTask = project.tasks.getByPath("detektMain") as Detekt
            val argumentString = detektTask.arguments.flatMap(CliArgument::toArgument).joinToString(" ")

            assertThat(argumentString).contains("--report xml:")
            assertThat(argumentString).contains("--report sarif:")
            assertThat(argumentString).doesNotContain("--report txt:")
        }
```

The DSL used won't exactly match what users will use, but for these tests I don't think that's an issue. `apply<KotlinPluginWrapper>()` is functionally identical to `apply("org.jetbrains.kotlin.jvm")`, and applying plugins using `apply<>()` is valid and [documented](https://docs.gradle.org/current/userguide/plugins.html#sec:applying_binary_plugins), even if it's not common. It's the same case for `configure<DetektExtension>`.

I wanted to check before migrating if there are any concerns or downsides I haven't considered. It's also possible to setup a project and use `.withProjectDir(File("/path/to/project"))` with ProjectBuilder so the tests that write files and then test the correct flags are passed to the CLI can still be tested.

**Comment by chao2zhang** [comment]
Integration tests with Gradle testkit are useful for testing Gradle groovy/kts DSL and would help build confidence for our Gradle plugins and extensions.

Meanwhile, I do agree that those integration tests with `dryRun` are not actually integration tests and should not lose test coverage once converted to unit tests written with `ProjectBuilder`. I would like to confirm the scope of this issue: It is only for those integration tests with dryRun option on. Is that true?

**Comment by 3flex** [comment]
> It is only for those integration tests with dryRun option on. Is that true?

Yes. Only the tests that are checking that the correct CLI flags are passed from the Gradle plugin, i.e. "dryRun = true".

Absolutely agree that we need functional tests for the reasons you mention, and they should continue using TestKit.

Terminology-wise I'm effectively saying that (in line with Gradle's [plugin testing terminology](https://docs.gradle.org/current/userguide/testing_gradle_plugins.html#tooling_support)) we would have standard unit tests (if applicable), integration tests that use ProjectBuilder, and functional tests that use TestKit.

**Comment by chao2zhang** [comment]
Today I learned about the [Gradle plugin testing support](https://docs.gradle.org/current/userguide/testing_gradle_plugins.html). I think we are aligned on the testing pyramid that we want to have sufficient unit test coverage but also verify with integration and functional test.

**Comment by cortinico** [comment]
> The "dry run" tests are just checking that when certain options are set on the plugin that the flags passed to the CLI are as expected. A slight refactoring in the plugin combined with switching to ProjectBuilder for the dry run tests will speed things up a lot - individual tests that take 25-30 seconds with TestKit take 15 with ProjectBuilder on my machine.

Definitely agree on this 👍 TestKit tests are extremely expensive.

I'm also afraid this might affected the experience of new contributors.
I believe we could consider isolating the integration tests to run only on CI or upon a specific task (i.e. not as dependency of the top level `build` command).

Mostly because if a contributor is interested in fixing a false positive, they probably are not interested in checking if the Gradle Plugin works fine on a Kotlin Multiplatform project ¯\_(ツ)_/¯

**Comment by BraisGabin** [comment]
So true. And that doesn't affect only *new* contributors. Maybe something as we have the warnings as errors in CI.

**Comment by cortinico** [comment]
Given that #4075 was merged, what else is left to do here?

**Comment by 3flex** [comment]
That was just one test. There other dry run tests should be migrated too.

**Comment by BraisGabin** [comment]
This is no longer an RFC, right? Can we change the title?
